README.md
# Headhesive.js

> An on-demand sticky header.

## What is it?

Headhesive.js creates an on-demand sticky header. Specify when you want your header to become fixed and the rest is magic. [View demo](http://markgoodyear.com/labs/headhesive/).

## Install

### Install with npm

```bash
npm install
```

### Install with yarn

```bash
yarn install headhesive
```

## How to use?

Headhesive.js is a standalone JavaScript plugin with no dependencies. Include the [`headhesive.mjs`](dist/headhesive.mjs) *(from the [`dist/`](dist/) folder)* in the footer of your page and initialise it:

```javascript
// Create a new instance of Headhesive
var header = new Headhesive('.header');
```

### Styling Headhesive.js

Headhesive.js doesn't inject any CSS styles so you can use your own to completely control how your sticky element behaves. An [example is provided in the demo](demo/css/headhesive.css).

## Options

Customise how Headhesive.js works by passing in custom options.

```javascript
// Options
var options = {
  offset: 500
}

// Create a new instance of Headhesive.js and pass in some options
var header = new Headhesive('.header', options);
```

### Defaults

```javascript
{
  // Scroll offset. Accepts Number or "String" (for class/ID)
  offset: 300, // OR — offset: '.classToActivateAt',

  // If using a DOM element, we can choose which side to use as offset (top|bottom)
  offsetSide: 'top',

  // Custom classes
  classes: {

    // Cloned elem class
    clone: 'headhesive',

    // Stick class
    stick: 'headhesive--stick',

    // Unstick class
    unstick: 'headhesive--unstick'
  },

  // Throttle scroll event to fire every 250ms to improve performace
  throttle: 250,

  // Callbacks
  onInit:    function () {},
  onStick:   function () {},
  onUnstick: function () {},
  onDestroy: function () {},
}
```

## Destroy method

To destroy an instance of Headhesive.js, you can call the destroy method:

```javascript
header.destroy();
```

## Browser support

IE9+ and modern browsers.

## License

Headesive.js is licensed under the [MIT License](LICENSE).

## Demo

There is a demo available in the [demo folder](demo/) and online at [http://markgoodyear.com/labs/headhesive/](http://markgoodyear.com/labs/headhesive/).


CHANGELOG.md
# Headhesive changelog

## v2.0.0

- Transitioned from Gulp to Vite build system.
- Refactored JavaScript codebase for improved efficiency and readability.
- Conducted backward compatibility testing.
- Updated the Readme file to reflect changes and provide up-to-date information.

## v1.2.4

- Fix resize event when using numerical offset
- Fix destroy method for resize event
- Changelog corrections

## v1.2.3

- Readme correction

## v1.2.2

- Fix UMD support

## v1.2.1

- Bump version to publish to npm

## v1.2.0

- Add UMD support
- Allow to select either the top or bottom of the offset element
- Update the offset position on browser resize
- Publish to npm
- Add license file
- Make the clone version lower z-index than original. Prevents visual overlap of both when scrolling quickly

## v1.1.1

- Fix class naming

## v1.1.0

- Call the init method internally on instance creation
- Destroy method to only destroy it's instance

## v1.0.0

- Initial


tests/spec/tests.js
(function(window) {
  "use strict";

  function isDeepEqual( test, target ) {
    var equal = true, key;

    for ( key in test ) {
      if ( test.hasOwnProperty( key ) ) {
        if ( JSON.stringify(test[key]) !== JSON.stringify(target[key]) ) {
          equal = false;
          break;
        }
      }
    }

    return equal;
  }

  describe( "Headhesive", function() {

    // setup default options from library to use for testing
    var defaultOptions = {
          offset: 300,
          classes: {
            clone: 'headhesive',
            stick: 'headhesive--stick',
            unstick: 'headhesive--unstick'
          },
          throttle: 250,
          onInit: function() {},
          onStick: function() {},
          onUnstick: function() {},
          onDestroy: function() {},
        },
      headhesive, fixture;

    beforeEach(function() {
      fixture = document.querySelector(".fixture");
    });

    afterEach(function() {
      if ( headhesive && headhesive.elem ) {
        headhesive.init(); //TODO: Shouldn't need to do this
        headhesive.destroy();
      }
    });

    describe( "Constructor", function() {
      it( "should exist", function() {
        expect( window.Headhesive ).toBeDefined();
      });

      describe("elem", function() {
        it("should store 1st parameter as an elem property", function() {
          headhesive = new Headhesive( fixture );

          expect( headhesive.elem ).toBe( fixture );
        });

        it("should find a DOM element if 1st parameter is a string", function() {
          headhesive = new Headhesive( ".fixture" );

          expect( headhesive.elem ).toBe( fixture );
        })
      });

      describe( "default options", function() {
        it( "should have a default set of options", function() {
          headhesive = new Headhesive( fixture );

          expect( isDeepEqual(headhesive.options, defaultOptions) ).toBe( true );
        });

        it( "should merge options with default options", function() {
          // set new options
          headhesive = new Headhesive( fixture, {
            throttle: 1000,
            classes: {
              stick: "new-stick-class"
            }
          });

          // set default options to match new options
          defaultOptions.throttle = 1000;
          defaultOptions.classes.stick = "new-stick-class"

          expect( headhesive.options.throttle ).toEqual( defaultOptions.throttle );
          expect( headhesive.options.classes.stick ).toEqual( defaultOptions.classes.stick );
        });
      });

    });

    describe( "Init", function() {
      it("should be defined", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.init ).toBeDefined();
      });

      it( "should make a clone of the passed element", function() {
        headhesive = new Headhesive( fixture );

        expect( fixture.cloneNode() ).toBeTruthy();
      });

      it( "should insert a cloned element into the DOM", function() {
        var countBeforeInit = document.body.childElementCount;

        headhesive = new Headhesive( fixture, {
          onInit: function() {
            return document.body.childElementCount
          }
        });

        expect(countBeforeInit).toBeLessThan(headhesive.options.onInit());
      });

      it("should invoke the onInit callback", function(){
        headhesive = new Headhesive( fixture, {
          onInit: function() {
            return true
          }
        });

        expect(headhesive.options.onInit()).toBeTruthy();
      });

      it("should add a scroll event listener on the window", function(){
        spyOn( window, "addEventListener" );

        headhesive = new Headhesive( fixture );

        expect( window.addEventListener ).toHaveBeenCalled();
      });

      //TODO: scrollOffset code could be extracted into its own method
      describe("scrollOffset", function() {
        it("should set the scrollOffset if option is a number", function() {
          headhesive = new Headhesive( fixture, {
            offset: 999
          });

          expect(headhesive.options.offset).toEqual(jasmine.any(Number));
        });

        it("should set the scrollOffset based off a selector if option is a string", function(){
          headhesive = new Headhesive( fixture, {
            offset: ".offset-test"
          });

          expect(headhesive.options.offset).toEqual(jasmine.any(String));
        });
      });

      describe("throttle", function() {
        it("should create a throttle method based on the option", function() {
          headhesive = new Headhesive( fixture, {
            throttle: 123
          });

          expect( headhesive._throttleUpdate ).toBeDefined();
        });

        it("should set the throttle to a different number than the default", function() {
          headhesive = new Headhesive( fixture, {
            throttle: 123
          });

          expect( isDeepEqual(headhesive.options, defaultOptions) ).toBe( false );
        });
      });
    });

    describe( "Destroy", function() {
      it("should be defined", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.destroy ).toBeDefined();
      });

      it("should invoke the onDestroy callback", function() {
        headhesive = new Headhesive( fixture, {
          onDestroy: function() {
            return true
          }
        });

        expect(headhesive.options.onDestroy()).toBeTruthy();
      });

      it("should remove elem from the DOM", function() {
        var countBeforeInit = document.body.childElementCount;

        headhesive = new Headhesive( fixture, {
          onDestroy: function() {
            return document.body.childElementCount
          }
        });

        expect(countBeforeInit).toBeLessThan(headhesive.options.onDestroy());
      });

      it("should remove scroll eventListener from the window", function(){
        spyOn( window, "removeEventListener" );

        headhesive = new Headhesive( fixture );
        headhesive.destroy();

        expect( window.removeEventListener ).toHaveBeenCalled();
      });
    });

    describe( "Stick", function() {
      it("should be defined", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.stick ).toBeDefined();
      });

      it("should set the stick class from default options", function() {
        headhesive = new Headhesive( fixture );

        expect( defaultOptions.classes.stick ).toEqual( jasmine.any(String) );
      });

      it("should set the stick class from options override", function() {
        headhesive = new Headhesive( fixture, {
          classes: {
            stick: "sample-stick-class"
          }
        });

        expect( isDeepEqual(headhesive.options, defaultOptions) ).toBe( false );
      });

      it("should invoke the onStick callback", function() {
        headhesive = new Headhesive( fixture, {
          onStick: function() {
            return true
          }
        });

        expect(headhesive.options.onStick()).toBeTruthy();
      });
    });

    describe( "Unstick", function() {
      it("should be defined", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.unstick ).toBeDefined();
      });

      it("should set the unstick class from default options", function() {
        headhesive = new Headhesive( fixture );

        expect( defaultOptions.classes.unstick ).toEqual( jasmine.any(String) );
      });

      it("should set the unstick class from options override", function() {
        headhesive = new Headhesive( fixture, {
          classes: {
            unstick: "sample-unstick-class"
          }
        });

        expect( isDeepEqual(headhesive.options, defaultOptions) ).toBe( false );
      });

      it("should invoke the onUnstick callback", function() {
        headhesive = new Headhesive( fixture, {
          onUnstick: function() {
            return true
          }
        });

        expect(headhesive.options.onUnstick()).toBeTruthy();
      });
    });

    describe( "Update", function() {
      it("should be defined", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.update ).toBeDefined();
      });

      it( "should create an update method", function() {
        headhesive = new Headhesive( fixture );

        expect( headhesive.update ).toBeDefined();
      });
    });
  });

}(window));


gulpfile.js
/**
 * Modules
 */
var gulp    = require('gulp');
var jshint  = require('gulp-jshint');
var concat  = require('gulp-concat');
var uglify  = require('gulp-uglify');
var header  = require('gulp-header');
var rename  = require('gulp-rename');
var include = require('gulp-include');
var bump    = require('gulp-bump');
var del     = require('del');
var pkg     = require('./package.json');

/**
 * Config
 */
var config = {
  name:   'Headhesive',
  src:    'src/headhesive.js',
  dest:   'dist/',
  banner: ['/*!',
           ' * Headhesive.js v<%= pkg.version %> - <%= pkg.description %>',
           ' * Author: Copyright (c) <%= pkg.author.name %> <<%= pkg.author.twitter %>> <<%= pkg.author.url %>>',
           ' * Url: <%= pkg.homepage %>',
           ' * License: <%= pkg.license %>',
           ' */',
           ''].join('\n')
};

/**
 * Bump version
 */
var bumpVer = function (bumpType) {
  return gulp.src(['./package.json', './bower.json'])
    .pipe(bump({
      type: bumpType
    }))
    .pipe(gulp.dest('./'));
}

gulp.task('bump:patch', function () { return bumpVer('patch'); })
gulp.task('bump:minor', function () { return bumpVer('minor'); })
gulp.task('bump:major', function () { return bumpVer('major'); })

/**
 * Tasks
 */
gulp.task('clean', function (cb) {
  del(['dist'], cb)
});

gulp.task('compile', function () {
  return gulp.src(config.src)
    .pipe(jshint('.jshintrc'))
    .pipe(jshint.reporter('jshint-stylish'))
    .pipe(include())
    .pipe(header(config.banner, { pkg: pkg }))
    .pipe(uglify({preserveComments: 'some', compress: false, mangle: false, output: { beautify: true, indent_level: 2 }}))
    .pipe(gulp.dest(config.dest))
    .pipe(uglify({preserveComments: 'some'}))
    .pipe(rename({suffix: '.min'}))
    .pipe(gulp.dest(config.dest));
});

gulp.task('default', ['clean'], function () {
  gulp.start('compile');
});

gulp.task('watch', function () {
  gulp.watch(config.src, ['compile']);
});


src/helpers.js
/**
 * Merge objects
 * @param  {Object} to
 * @param  {Object} from
 * @return {Object}
 */
export const mergeObj = function (to, from) {
  for (let p in from) {
    if (Object.prototype.hasOwnProperty.call(from, p)) {
      to[p] = typeof from[p] === "object" ? mergeObj(to[p], from[p]) : from[p];
    }
  }

  return to;
};

/**
 * Throttle, borrowed from Underscore.js
 */
export const throttle = function (func, wait) {
  const now =
    Date.now ||
    function () {
      return new Date().getTime();
    };
  let context, args, result;
  let timeout = null;
  let previous = 0;
  const later = function () {
    previous = now();
    timeout = null;
    result = func.apply(context, args);
    context = args = null;
  };

  return function () {
    const now = Date.now();
    const remaining = wait - (now - previous);
    context = this;
    args = arguments;

    if (remaining <= 0) {
      clearTimeout(timeout);
      timeout = null;
      previous = now;
      result = func.apply(context, args);
      context = args = null;
    } else if (!timeout) {
      timeout = setTimeout(later, remaining);
    }

    return result;
  };
};

/**
 * Get current Y position
 * @return {Number}
 */
export const getScrollY = function () {
  return window.scrollY !== undefined
    ? window.scrollY
    : (document.documentElement || document.body.parentNode || document.body)
        .scrollTop;
};

/**
 * Get elements Y position
 * @param  {Object | Number} elem - The element
 * @param  {String}          side - Elem side (top or bottom)
 * @return {Number}
 */
export const getElemY = function (elem, side) {
  let pos = 0;
  const elemHeight = elem.offsetHeight;

  while (elem) {
    pos += elem.offsetTop;
    elem = elem.offsetParent;
  }

  if (side === "bottom") {
    pos = pos + elemHeight;
  }

  return pos;
};


src/headhesive.js
import { mergeObj, throttle, getScrollY, getElemY } from "./helpers.js";

(function (root, factory) {
  if (typeof define === "function" && define.amd) {
    define([], function () {
      return factory();
    });
  } else if (typeof exports === "object") {
    module.exports = factory();
  } else {
    root.Headhesive = factory();
  }
})(this, function () {
  "use strict";

  var Headhesive = function (elem, options) {
    if (!("querySelector" in document && "addEventListener" in window)) {
      return;
    }

    this.visible = false;

    this.options = {
      offset: 300,
      offsetSide: "top",
      classes: {
        clone: "headhesive",
        stick: "headhesive--stick",
        unstick: "headhesive--unstick",
      },
      throttle: 250,
      onInit: function () {},
      onStick: function () {},
      onUnstick: function () {},
      onDestroy: function () {},
    };

    this.elem = typeof elem === "string" ? document.querySelector(elem) : elem;
    this.options = mergeObj(this.options, options);
    this.init();
  };

  Headhesive.prototype = {
    constructor: Headhesive,

    init: function () {
      this.clonedElem = this.elem.cloneNode(true);
      this.clonedElem.className += " " + this.options.classes.clone;
      document.body.insertBefore(this.clonedElem, document.body.firstChild);

      if (typeof this.options.offset === "number") {
        this.scrollOffset = this.options.offset;
      } else if (typeof this.options.offset === "string") {
        this.setScrollOffset();
      } else {
        throw new Error("Invalid offset: " + this.options.offset);
      }

      this.throttleUpdate = throttle(
        this.update.bind(this),
        this.options.throttle
      );
      this.throttleScrollOffset = throttle(
        this.setScrollOffset.bind(this),
        this.options.throttle
      );

      window.addEventListener("scroll", this.throttleUpdate, false);
      window.addEventListener("resize", this.throttleScrollOffset, false);
      this.options.onInit.call(this);
    },

    setScrollOffset: function () {
      if (typeof this.options.offset === "string") {
        this.scrollOffset = getElemY(
          document.querySelector(this.options.offset),
          this.options.offsetSide
        );
      }
    },

    destroy: function () {
      document.body.removeChild(this.clonedElem);
      window.removeEventListener("scroll", this.throttleUpdate);
      window.removeEventListener("resize", this.throttleScrollOffset);
      this.options.onDestroy.call(this);
    },

    stick: function () {
      if (!this.visible) {
        this.clonedElem.className = this.clonedElem.className.replace(
          new RegExp(
            "(^|\\s)*" + this.options.classes.unstick + "(\\s|$)*",
            "g"
          ),
          ""
        );
        this.clonedElem.className += " " + this.options.classes.stick;
        this.visible = true;
        this.options.onStick.call(this);
      }
    },

    unstick: function () {
      if (this.visible) {
        this.clonedElem.className = this.clonedElem.className.replace(
          new RegExp("(^|\\s)*" + this.options.classes.stick + "(\\s|$)*", "g"),
          ""
        );
        this.clonedElem.className += " " + this.options.classes.unstick;
        this.visible = false;
        this.options.onUnstick.call(this);
      }
    },
    update: function () {
      if (getScrollY() > this.scrollOffset) {
        this.stick();
      } else {
        this.unstick();
      }
    },
  };

  return Headhesive;
});
